File: sdk/typescript/src/client/FarsightRpcClient.ts
import { DEFAULT_API_URL } from "../constants.js";
import { FarsightRpcHttpError } from "../errors/FarsightRpcHttpError.js";
import { buildHeaders, resolveGlobalFetch, safeReadText } from "../internal/fetch.js";
import { mapGetRpcsResponse } from "../internal/mapGetRpcsResponse.js";
import type { FarsightRpcClientOptions, GetRpcsOptions, GetRpcsResult } from "../types.js";

export class FarsightRpcClient {
  private readonly apiUrl: string | URL;
  private readonly apiKey?: string;
  private readonly fetchImplementation;

  public constructor(options: FarsightRpcClientOptions = {}) {
    this.apiUrl = options.apiUrl ?? DEFAULT_API_URL;
    this.apiKey = options.apiKey ?? undefined;
    this.fetchImplementation = options.fetch ?? resolveGlobalFetch();
  }

  public async getRpcs(options: GetRpcsOptions = {}): Promise<GetRpcsResult> {
    const response = await this.fetchImplementation(new URL("/api/Rpcs", this.apiUrl), {
      headers: buildHeaders(this.apiKey),
      signal: options.signal,
    });

    switch(response.status) {
      case 403:
        return { kind: "invalid-api-key" };
      case 200:
        return mapGetRpcsResponse(await response.json());
      default:
        throw new FarsightRpcHttpError(response.status, await safeReadText(response));
    }
  }
}

File: sdk/csharp/Client/FarsightRpcClient.cs
using Farsight.Chains;
using Farsight.Rpc.Types;
using System.Collections.Immutable;
using System.Net;
using System.Net.Http.Json;
using static Farsight.Rpc.Sdk.Client.IFarsightRpcClient;

namespace Farsight.Rpc.Sdk.Client;

public sealed class FarsightRpcClient : IFarsightRpcClient
{
    private readonly IHttpClientFactory? _httpClientFactory;
    private readonly HttpClient? _httpClient;

    internal FarsightRpcClient(IHttpClientFactory httpClientFactory)
    {
        ArgumentNullException.ThrowIfNull(httpClientFactory);
        _httpClientFactory = httpClientFactory;
    }

    public FarsightRpcClient(FarsightRpcOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);
        _httpClient = new HttpClient();
        ConfigureClient(_httpClient, options);
    }

    public async Task<GetRpcsResult> GetRpcsAsync(string apiKey, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(apiKey);

        using var request = new HttpRequestMessage(HttpMethod.Get, "/api/Rpcs");
        request.Headers.Add(ApiKeyHeaders.API_KEY, apiKey);
        using var response = await CreateClient().SendAsync(request, cancellationToken);

        switch(response.StatusCode)
        {
            case HttpStatusCode.Forbidden:
                return GetRpcsResult.InvalidApiKey.Instance;
            case HttpStatusCode.OK:
                var result = await response.Content.ReadFromJsonAsync(FarsightRpcJsonContext.Default.ApiKeyRpcsDto, cancellationToken)
                    ?? throw new InvalidOperationException("Null response");
                var resolveProvider = (Guid providerId) => result.Providers.FirstOrDefault(x => x.Id == providerId)
                    ?? throw new InvalidOperationException($"RPC response referenced unknown provider '{providerId}'.");

                var rpcs = result.Rpcs.ToDictionary(
                    group => ChainRegistry.Chains.FirstOrDefault(x => x.Name == group.Key)
                        ?? throw new InvalidOperationException($"RPC response referenced unknown chain '{group.Key}'."),
                    group => group.Value.Select(rpc => new RpcEndpoint
                    {
                        Id = rpc.Id,
                        Address = rpc.Address,
                        Provider = resolveProvider(rpc.ProviderId),
                        Capabilities = rpc.Capabilities,
                        EthGetLogsLimit = rpc.EthGetLogsLimit,
                        Order = rpc.Order,
                    }).ToImmutableArray()
                );
                var publicRpcs = result.PublicRpcs.ToDictionary(
                    group => ChainRegistry.Chains.FirstOrDefault(x => x.Name == group.Key)
                        ?? throw new InvalidOperationException($"RPC response referenced unknown chain '{group.Key}'."),
                    group => group.Value
                );

                return new GetRpcsResult.Success(rpcs, publicRpcs, result.PublicRpcsUpdatedAt, result.Providers, result.ErrorGroups);
            default:
                response.EnsureSuccessStatusCode();
                throw new InvalidOperationException();
        }
    }

    private HttpClient CreateClient()
        => _httpClientFactory?.CreateClient(DependencyInjection.HTTP_CLIENT_NAME) ?? _httpClient!;

    internal static void ConfigureClient(HttpClient client, FarsightRpcOptions options)
    {
        client.BaseAddress = options.ApiUrl;
    }
}
